// data-provider/dist/LoadingComponent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoadingComponent = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LoadingComponent = exports.LoadingComponent = function LoadingComponent(_ref) {
  var _ref$color = _ref.color,
      color = _ref$color === undefined ? '#777' : _ref$color;
  return _react2.default.createElement(
    'div',
    { className: 'loader', title: '1' },
    _react2.default.createElement(
      'svg',
      { version: '1.1', id: 'loader-1', xmlns: 'http://www.w3.org/2000/svg', xmlnsXlink: 'http://www.w3.org/1999/xlink',
        x: '0px', y: '0px', width: '40px', height: '40px', viewBox: '0 0 50 50', xmlSpace: 'preserve'
      },
      _react2.default.createElement(
        'path',
        { fill: color,
          d: 'M25.251,6.461c-10.318,0-18.683,8.365-18.683,18.683h4.068c0-8.071,6.543-14.615,14.615-14.615V6.461z'
        },
        _react2.default.createElement('animateTransform', { attributeType: 'xml',
          attributeName: 'transform',
          type: 'rotate',
          from: '0 25 25',
          to: '360 25 25',
          dur: '0.6s',
          repeatCount: 'indefinite'
        })
      )
    )
  );
};
